File: HW6/hw6.java
import java.util.*;

public class hw6 {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums); // sort the nums to avoid dupes
        List<List<Integer>> result = new ArrayList<>();
        
        for (int i = 0; i < nums.length - 2; i++) {
            // skip dupes
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            
            int left = i + 1;
            int right = nums.length - 1;
            
            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (sum == 0) {
                    result.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    left++;
                    right--;
                    
                    // skip dupes
                    while (left < right && nums[left] == nums[left - 1]) {
                        left++;
                    }
                    while (left < right && nums[right] == nums[right + 1]) {
                        right--;
                    }
                } else if (sum < 0) {
                    left++;
                } else {
                    right--;
                }
            }
        }
        
        return result;
    }

    public static void main(String[] args){
        hw6 test = new hw6();
        int[] nums = {0,1,1};
        List<List<Integer>> result = test.threeSum(nums);
        System.out.println(result);
    }
}

File: HW6/hw6.py
from typing import List

class hw6():
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort() 
        result = []

        for i in range(len(nums) - 2):
            # skip dupes
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            left, right = i + 1, len(nums) - 1
            while left < right:
                current_sum = nums[i] + nums[left] + nums[right]
                if current_sum == 0:
                    result.append([nums[i], nums[left], nums[right]])
                    left += 1
                    right -= 1

                    # skip dupes
                    while left < right and nums[left] == nums[left - 1]:
                        left += 1
                    while left < right and nums[right] == nums[right + 1]:
                        right -= 1
                elif current_sum < 0:
                    left += 1  # needs larger num
                else:
                    right -= 1  # needs smaller num

        return result
    
# test = hw6()
# #n = [10,-10,0,0,20, -20, -5, -15]
# n = [0, 1, 1]
# result = test.threeSum(n)

# print(result)
